lib/staccato/V4/event.rb

```diff
@@ -0,0 +1,105 @@
+module Staccato::V4
+  module Event
+    def self.events
+      @events ||= {}
+    end
+
+    # this module is included into each model hit type
+    #   to share the common behavior required to hit
+    #   the Google Analytics /mp/collect api endpoint
+    def self.included(event)
+      event.extend Forwardable
+
+      event.class_eval do
+        attr_accessor :tracker, :options
+
+        def_delegators :@options, *event::FIELDS
+
+        # class name, underscored, symbolized
+        def self.event_name
+          @event_name || begin
+            # gsub!(/::/, '/').
+            self.
+              to_s.
+              dup.
+              split("::").
+              last.
+              gsub(/([A-Z]+)([A-Z][a-z])/,'\1_\2').
+              gsub(/([a-z\d])([A-Z])/,'\1_\2').
+              tr("-", "_").
+              downcase.
+              to_sym
+          end
+        end
+      end
+
+      def name
+        self.class.event_name
+      end
+
+      self.events[event.event_name] = event
+    end
+
+    def initialize(tracker, options = {})
+      self.tracker = tracker
+      self.options = Staccato::OptionSet.new(options)
+    end
+
+    # return the fields for this hit type
+    # @return [Array] the fields
+    def fields
+      self.class::FIELDS
+    end
+
+    def as_json
```

Comment:
I'm doing some digging to make sure this actually resolves in some fashion. Right now, I'm not finding it grepping the code.

Update: This convention is used within Ruby's JSON library and this is exactly what's needed for it. https://stackoverflow.com/questions/3226054/how-to-convert-a-ruby-object-to-json